### ISSUE TITLE
Sprint: April 6 → April 18

### ISSUE THREAD

**Issue by RichardLitt** [ipfs/pm#101]
# Sprint April 6, 2016

## Sprint Discussions

[Schedule](https://github.com/ipfs/pm#sprint-discussion-schedule)

Please take notes in a separate pad, if you can, and link it here.

Endeavour         | Lead            | Time (PDT - **UTC/Z** - CET) | Pad
:---------------: | :-------------: | :--------------------------: | :----:
sync              | @RichardLitt    | 9:00PDT **16:00Z** 18:00CET  | IRC: #ipfs on Freenode
apps on ipfs      | @dignifiedquire | 10:30PDT **17:30Z** 19:30CET | https://public.etherpad-mozilla.org/p/ipfs-april-6-apps-on-ipfs
infrastructure    | @lgierth        | 11:00PDT **18:00Z** 20:00CET | https://public.etherpad-mozilla.org/p/ipfs-april-6-infrastructure
libp2p            | @diasdavid      | 11:30PDT **18:30Z** 20:30CET | https://public.etherpad-mozilla.org/p/ipfs-april-6-libp2p
js-ipfs           | @diasdavid      | 12:00PDT **19:00Z** 21:00CET | https://public.etherpad-mozilla.org/p/ipfs-april-6-js-ipfs
go-ipfs           | @whyrusleeping  | 12:30PDT **19:30Z** 21:30CET | https://public.etherpad-mozilla.org/p/ipfs-april-6-go-ipfs

Please add the Agenda to the Pad before the endeavour sprint starts.

## Sprint Deliverables

- Add your tasks below in a comment, that way we only have people listed who are really in the sprint
- Add links to issues down here. Only add things you can finish this sprint.

**Comment by dignifiedquire** [comment]
## Apps on IPFS
#### Lead: @dignifiedquire

#### Participants

- @dignifiedquire
- @noffle
- @richardlitt

## Agenda

- webui update
- dynamic apps on ipfs
  - ipfs-log (https://github.com/haadcode/ipfs-log/)
  - pubsub
  - CRDTs
  - Orbit and IPFS-Pad (https://github.com/haadcode/orbit)

## Notes

### Webui Updates

- Ship Redux PR with all new features soon
- Wait for user feedback

### Dynamic Apps on IPFS

- ipfs-log (https://github.com/haadcode/ipfs-log/)
  - Is an CRDT with causal linking
  - Wants PubSub: (Near) realtime updates

### Questions:

**mmuller**: Does a pubsub mechanism currently exist?
**Answer**: No, prototype in go-ipfs atm. And there is orbit-db which uses redis as a shim for now

**vijayee**: does the crdt code exist already?
**Answer**: Partially, noffle implemented ipfs-log but it's not fully compliant with all requirements for a CRDT yet.

**Comment by lgierth** [comment]
## Infrastructure

#### Lead: @lgierth

#### Participants

- @jbenet 
- @lgierth

## Agenda

- last-minute v0.4.0 change
  - extend v03x/multireq EOL from end-of-april to end-of-june-ish
    - measure v03x vs. v04x servings
    - looks at decreases in v03x nodes
- dns post mortem & actionables
  - codify dns zones
  - tool to replay the zone files to different dns providers
  - multiple dns providers
  - fallback procedure
- metrics / tests

**Comment by diasdavid** [comment]
## libp2p

#### Lead: @diasdavid

## Notes

- update
- next steps
  - finish integration of the new browserify-zlib into libp2p-spdy
  - update libp2p-spdy on libp2p-swarm and enable browser tests 
  - update libp2p-ipfs to use the latest libp2p-swarm and libp2p-spdy
  - update js-ipfs to use latest libp2p-ipfs

**Comment by diasdavid** [comment]
## js-ipfs

#### Lead: @diasdavid

## Notes

- update
- we're close™
- captain.log
  - https://github.com/ipfs/js-ipfs/issues/30
- contribute
  - https://github.com/ipfs/js-ipfs/#contribute
- Files API by Nathan
- bitswap

**Comment by diasdavid** [comment]
### js-ipfs

- **management**
- **jsipfs swarm**
  - [ ] peers
  - [ ] addrs
  - [ ] connect
  - [ ] disconnect
  - [ ] interop
- **js-libp2p-ipfs**
- **js-ipfs-data-importing**

### js-ipfs-api

### libp2p

- **webrtc-explorer**
- **js-libp2p-swarm**
- **js-libp2p-tcp**
- **js-libp2p-utp**
- **js-libp2p-websockets**
- **js-libp2p-multiplex**

### npm on IPFS

### IPLD

### webui

### specs

- **http-api-spec**
- **specs**
  - [ ] get core api spec with @jbenet  
  - [ ] get unixfs spec with @jbenet & @whyrusleeping 
  - [ ] get Files API spec with @jbenet & @whyrusleeping 

### extra

**Comment by whyrusleeping** [comment]
# go-ipfs

Lead: @whyrusleeping

# Notes

- talked about Q2 goals
- talked about shipping 0.4.0
  - need one or two more changes
  - 0.4.0-dev will not be compatible with 0.4.0 release
- file descriptor limit raising code
  - add into 0.4.0 release
  - enable with flag

**Comment by RichardLitt** [comment]
### To Do
- [ ] Sprint Management
- [ ] Write the Weekly roundup
- [ ] Take a look at the ipfs.pics blogpost https://github.com/ipfs/blog/pull/15
- [ ] Review this PR from David https://github.com/ipfs/community/pull/111

**Comment by dignifiedquire** [comment]
## ToDo

- [x] Release dignified.js@1.0.0
- [ ] Bring js-ipld up to date with the spec
- [ ] IPLD playground preparations 
- [ ] Learning go
- [ ] Small go-ipfs PRs to get my feet wet
  - [ ] https://github.com/ipfs/go-ipfs/pull/2526
  - [ ] https://github.com/ipfs/go-ipfs/issues/2456

**Comment by nginnever** [comment]
<h2>t0Do</h2>
- dignified.js
 - [ ] CR https://github.com/jbenet/js-multihash/pull/5
 - [ ] CR https://github.com/diasdavid/js-ipld/pull/9
 - [ ] CR https://github.com/diasdavid/js-libp2p-distributed-record-store/pull/3
- js-ipfs-data-importing
 - [~] exporter
 - [ ] Rework importer
- js-ipfs
 - [~] cat command
 - [~] get command
 - [ ] ls command
 - [~] add command
- Webui
 - [ ] write tests
 - [ ] document in contibuting.md

**Comment by noffle** [comment]
## noffle

- go-ipfs
 - [ ] CLI docs: discoverability & consistency (https://github.com/ipfs/go-ipfs/pull/2542)
 - [ ] output directories when 'ipfs add -rn' is used (https://github.com/ipfs/go-ipfs/pull/2531)
- ipfs blog
 - [ ] explain the power of the Files API more clearly (https://github.com/ipfs/blog/pull/43)
- js-ipfs
 - [ ] tackle blocking Milestone 1 issues
- ipfs-pad
 - [ ] write String CRDT
 - [ ] rig up a prototype front-end and get something working in the browser

**Comment by haadcode** [comment]
### @haadcode sprint goals - 6 Apr - 18 Apr

#### Summary
With the release of 0.4, I'll be upgrading orbit-related modules to use 0.4. I will continue to work on orbit-db and ipfs-log, fixing some issues, optimizing the algorithms and working towards to make them usable in the browser and better facilitate to build the CRDT library on top of ipfs-log. I also want to fix some issues in Orbit, and work towards making it usable in browser-only (stand-alone webapp).

#### Tasks

##### orbit-db
  - [ ] Upgrade to IPFS 0.4
  - [ ] Stretch goal: Add pinning to db operations and ipfs-log objects

##### ipfs-log
  - [ ] Upgrade to IPFS 0.4
  - [ ] optimize findHeads
  - [ ] fix sorting on merge to be deterministic
  - [ ] Stretch goal: Stream API for ipfs-log

##### Orbit
  - [ ] Upgrade to IPFS 0.4
  - [ ] Button for copying the IPFS hash of each message
  - [ ] Sort messages by timestamps instead of log order
  - [ ] Stretch goal: Multiline POST

**Comment by lgierth** [comment]
### @lgierth todo

- go-ipfs
  - [x] release 0.4.0 :ship:
  - [ ] metrics: rework prometheus integration ipfs/go-ipfs#2356
  - [ ] repo: properly init Datastore config, and leave it be ipfs/go-ipfs#2549
  - [ ] go-libp2p: fix address filtering for /ip6, add tests ipfs/go-libp2p#36 
  - [ ] readme: mention building from source on windows ipfs/go-ipfs#2550
- infrastructure
  - [x] 0.4.0 release outfall: disable utp listeners
  - [x] 0.4.0 release outfall: disable ssh authorized_keys automation
  - [ ] revert these adhoc workarounds
  - [x] prevent gateways from dialing each other over cjdns (attempt at fixing slowness) ipfs/infrastructure@b34f10337cc3c27c35637d49633e8e4f1dd58a5a
  - [ ] postmortem for trailing-comma outage on april 9th
- DNS outage actionables
  - [ ] merge postmortem ipfs/ops-private#4
  - [ ] codify domains
  - [ ] prepare moving domains to gandi company account
  - [ ] better ipns tests
  - [ ] lower ttl for all ipfs.io zones
  - [ ] postmortem blog post
- org things
  - [x] Content-Type of the Week ipfs/weekly#34
  - [x] add myself to the people directory protocol/main#17
  - [ ] fly to Seattle
  - [ ] fly to Toronto

**Comment by whyrusleeping** [comment]
@whyrusleeping TODO

- [x] fix bootstrapping issue
- [x] make quick hack of gx pinning website
- [ ] websocket transport
  - [ ] decide how websocket multiaddrs should work
- [ ] gxify libp2p-transport (and deps)
  - [ ] readmes and CI for new repos
- [ ] unixfs sharding
  - waiting on @jbenet for spec
- [ ] maybe some pubsub work

**Comment by dignifiedquire** [comment]
## ToDo

- [ ] Fix browserify-zlib finally https://github.com/ipfs/browserify-zlib/pull/2 to unblock @diasdavid 
- [ ] Bring js-ipld up to date with the spec
- [ ] ipld.io
  - [ ] IPLD playground 
  - [ ] render spec
- [ ] Learning go
- [ ] Small go-ipfs PRs to get my feet wet
  - [ ] https://github.com/ipfs/go-ipfs/pull/2526
  - [ ] https://github.com/ipfs/go-ipfs/issues/2456

**Comment by diasdavid** [comment]
## David Dias Check In

> **Summary:** This sprint has been intense (and it will continue), with the goal of getting everything in line for a beta version of `js-ipfs` by next week. Lot's of progress has been made across the panoply of JavaScript modules, from more test coverage, integration of the aegir assets pipeline, better documentation, project management, new features for `jsipfs swarm`, namely: connect, peers and addrs local. My attention and time was across the board, CR+Merging all the things, help unblock things, get zlib integrated and (finally) spdy enabled over WebSockets (with flying colours).

**Comment by haadcode** [comment]
### @haadcode sprint end - 6 Apr - 18 Apr

#### Summary
I concentrated on upgrading all my projects to use 0.4, getting orbit-db and ipfs-log to work in the browser and improving Orbit's UX. Had a very productive sprint. I managed to upgrade everything to 0.4 and it's awesome: much faster and more stable than 0.3! We now have working distributables for **orbit-db and ipfs-log in the browser**. They still require a local daemon to run to work but this will change when js-ipfs ships. I added some new features to Orbit: preview files directly in the chat with code higlighting, players for audio and video and improved the files browsing functionality in general. You can now also copy the hash of a file to clipboard. It looks like this:

<img src="https://raw.githubusercontent.com/haadcode/orbit/master/screenshots/screenshot6%202016-04-17.png" height="320">

I also created a simple javascript logging module called [logplease](https://github.com/haadcode/logplease), works in Node.js and browsers.

#### Tasks

##### orbit-db
  - [ ] Upgrade to IPFS 0.4
  - [ ] Make it work in the browser
  - [ ] Stretch goal: Add pinning to db operations and ipfs-log objects

##### ipfs-log
  - [ ] Upgrade to IPFS 0.4
  - [ ] Make it work in the browser
  - [ ] optimize findHeads
  - [ ] fix sorting on merge to be deterministic
  - [ ] Stretch goal: Stream API for ipfs-log

##### Orbit
  - [ ] Upgrade to IPFS 0.4
  - [ ] Button for copying the IPFS hash of each message
  - [ ] Sort messages by timestamps instead of log order
  - [ ] Bugfixing
  - [ ] Extra: preview files inside Orbit
  - [ ] Stretch goal: Multiline POST

**Comment by lgierth** [comment]
### @lgierth sprint update

One and a half weeks ago we released go-ipfs 0.4.0, yay! I took the smaller part
in this intense day, doing dist.ipfs.io and go-ipfs deploys. I did some more
work on go-ipfs and go-libp2p during the rest of the sprint -- a couple of
bug fixes, and dependency updates. I'm now much more comfortable with gx.

Mid-sprint I flew to Seattle, which is disturbingly warm and green at the moment.
I caught a bad cold right away, and didn't get much more done. I also had the
chance to meet and hang out with @thefinn93 and @danry25 of Seattle Meshnet. We've been
working on cjdns and OpenWrt for two years already, but have never met before :)

This morning I landed in Toronto, where I'll spend most of the week working on
the cjdns android app with @benhylau -- IPFS work will happen too, but not too
much.

Task list:

- go-ipfs
  - [x] release 0.4.0 :ship:
  - [x] metrics: rework prometheus integration ipfs/go-ipfs#2356
  - [x] repo: properly init Datastore config, and leave it be ipfs/go-ipfs#2549
  - [ ] backport and deploy ipfs/go-ipfs#2549 to v03x ipfs/go-ipfs@0-3-stable
  - [x] repo: don't create logs directory, not used any longer ipfs/go-ipfs#2569
  - [x] go-libp2p: fix address filtering for /ip6, add tests ipfs/go-libp2p#36
  - [x] go-libp2p: extract stuff into its own packages ipfs/go-libp2p#40
  - [ ] RFM update go-libp2p dependency ipfs/go-ipfs#2576
  - [ ] readme: mention building from source on windows ipfs/go-ipfs#2550
- infrastructure
  - [x] 0.4.0 release outfall: disable utp listeners
  - [x] 0.4.0 release outfall: disable ssh authorized_keys automation
  - [x] revert these adhoc workarounds
  - [ ] waiting on libp2p: prevent gateways from dialing each other over cjdns (attempt at fixing slowness) ipfs/infrastructure@b34f10337cc3c27c35637d49633e8e4f1dd58a5a
  - [x] postmortem for trailing-comma outage on april 9th ipfs/ops-private#6
  - [x] metrics: invert "services online" graph, better signal/noise ratio http://metrics.i.ipfs.io/dashboard/db/meta
  - [ ] metrics: fix mtail's duplicate output
- DNS outage actionables
  - [x] merge postmortem ipfs/ops-private#4
  - [ ] codify domains
  - [ ] prepare moving domains to gandi company account
  - [ ] better ipns tests
  - [ ] lower ttl for all ipfs.io zones
  - [ ] postmortem blog post
- org things
  - [x] Content-Type of the Week ipfs/weekly#34
  - [x] add myself to the people directory protocol/main#17
  - [x] fly to Seattle
  - [x] fly to Toronto

**Comment by nginnever** [comment]
<h3>@nginnever Check-In</h3>

This past week I spent more time working with data-exporting, mostly testing moving files back and forth from go-ipfs to js-ipfs and hunting down some bugs. This lead to studying more about node.js streams and having conversations with dignfiedquire and daviddias. Big thanks to them for helping me understand so much about streaming data. Also started learning reactive js to try out a module called highland by dignifiedquire's suggestions. This module looks powerful in its way of simplifying async code and I plan to incorporate it once I understand a bit more about reactive js programming. 

I spent a minimal amount of time working with aegir, the testing suit and standards for js modules. Reworked the multihash module to update from dig.js to aegir. 

Over the weekend I met up with alu in hollywood where we discussed visions of the future and how the web might look and be navigated. His work with janus VR and the equipment he has is really cool.  Hollywood shows a lot of interest in visual things so it's natural to be asking these questions there. Seeing his work flow I started to come up with ideas of how IPFS might optimize data traffic for VR which could potentially create an interactive web that is content addressed in a way where you can actually walk through a dag of linked rooms! 

- aegir
 - [x] PR https://github.com/jbenet/js-multihash/pull/5
 - [x] PR https://github.com/diasdavid/js-ipld/pull/9
 - [x] PR https://github.com/diasdavid/js-libp2p-distributed-record-store/pull/3
- js-ipfs-data-importing
 - [~] exporter
 - [ ] Rework importer
- js-ipfs
 - [~] cat command
 - [~] get command
 - [ ] ls command
 - [~] add command
- Webui
 - [ ] write tests
 - [ ] document in contibuting.md
- go-ipfs
 - [ ] study go-lang

**Comment by whyrusleeping** [comment]
### @whyrusleeping sprint update
Started off last week by fixing the libp2p listener bug that snuck into 0.4.0.  After that, I set to work trying to get enough pull requests merged that we would have under 40 open. Unfortunately we kept opening new ones at a rate higher than I could get through them, so its still nearly 50. I also started modularizing the go side of the codebase, breaking a few packages out of go-libp2p, @lgierth has a PR that brings this into go-ipfs now too. In the same mood, I went and removed a tonne and a half of unused godeps, the go-ipfs codebase is a tad lighter now.

This week i'm hoping to refactor the swarm dialer in go-libp2p to allow better address intake control, and hopefully fix the dht findpeer bug at the same time.

**Comment by noffle** [comment]
## @noffle end-of-sprint mind dump

- go-ipfs
 - [~] CLI docs: discoverability & consistency (https://github.com/ipfs/go-ipfs/pull/2542)
 - [~] output directories when 'ipfs add -rn' is used (https://github.com/ipfs/go-ipfs/pull/2531)
 - [~] human-/tool-friendly 'ipfs dht' output (https://github.com/ipfs/go-ipfs/issues/2494)
 - [x] finish @eminence's "Adds tests to make sure 'object patch' writes" PR (https://github.com/ipfs/go-ipfs/pull/2555)
- ipfs blog
 - [x] explain the Files API a bit more clearly (https://github.com/ipfs/blog/pull/43)
- js-ipfs
 - [x] tackle blocking Milestone 1 issues
 - [x] js-peer-id: 100% test coverage, README, CircleCI
 - [x] js-mulitaddr: README
 - [x] js-peer-info: 100% test coverage, README, CircleCI
 - [x] js-ipfs-blocks: 100% coverage, README

**Comment by dignifiedquire** [comment]
## @dignifiedquire sync update

- [x] Fix browserify-zlib finally https://github.com/ipfs/browserify-zlib/pull/2 to unblock @diasdavid 
- [x] dignified.js/aegir
  - [x] renamed to aegir
  - [x] added contributor update task
  - [x] setup on most repos https://github.com/dignifiedquire/aegir/issues/1 
- [ ] IPLD
  - [x] Implement [js-ipfs-ipld](https://github.com/ipfs/js-ipfs-ipld)
  - [x] Update [js-ipld](https://github.com/ipfs/js-ipld) to the current spec
  - [x] Implement [js-ipld-cli](https://github.com/ipfs/js-ipld-cli)
  - [~] read discussions
  - [~] ipld.io
    - [~] IPLD playground 
    - [~] render spec
- [ ] Learning go
- [ ] Small go-ipfs PRs to get my feet wet
  - [ ] https://github.com/ipfs/go-ipfs/pull/2526
  - [ ] https://github.com/ipfs/go-ipfs/issues/2456

**Comment by RichardLitt** [comment]
### Update

This week, I spent a few days doing cool stuff: I cleared out most of my open issues, and merged a few of the http-api pulls. I worked really hard on the js guidelines for IPFS. I also looked into readme-standard again, learned how to make remark-lint plugins (and [created remark-lint-appropriate-heading](https://github.com/RichardLitt/remark-lint-appropriate-heading), the first of these needed). I also worked on the weekly (which still hasn't been published, I think in a month now). I need to figure out how to make this work, because right now it is still a very broken pipeline. Also, I got horrible food poisoning. 

- [x] Sprint Management
- [x] Write the Weekly roundup
- [~] Merge js-ipfs guidelines https://github.com/ipfs/community/pull/111
- [ ] Take a look at the ipfs.pics blogpost https://github.com/ipfs/blog/pull/15
- [x] Closed http-api-spec `pin` https://github.com/ipfs/http-api-spec/pull/48
- [x] Updated http-api-spec `dht` https://github.com/ipfs/http-api-spec/pull/44
- [x] Updated go-ipfs-api readme a bit https://github.com/ipfs/go-ipfs-api

**Comment by jbenet** [comment]
**Summary**
I was focused on: shipping 0.4.0 and all the things around it,  helping find/address bootstrap hangs (fixed in 0.4.1), lots of discussions with @diasdavid and @whyrusleeping for go/js interop, IPLD discussions (in person and https://github.com/ipfs/specs/issues/91), testing Orbit (so cool!), meetings/discussions about ipfs with other orgs (ob, gitlab, lantern), and upkeep on lots of ongoing conversations with other orgs.

**Some things done**

- [x] shipping 0.4.0 and all the things around it
  - [x] review blog post
  - [x] update website mentions
  - [x] test ipfs installation from various paths
  - [x] test + found problems in ipfs-update
  - [x] all the distributions republishing
  - [x] multistream fixups
- [x] helping find/address bootstrap hangs (fixed in 0.4.1)
  - [x] bootstrapper rebooter 
- [x] lots of discussions with @diasdavid and @whyrusleeping for go/js interop
  - [x] sharding plans (todo: write up)
  - [x] bitswap spec and walkthrough
- [x] IPLD discussions with @nicola (ongoing here https://github.com/ipfs/specs/issues/91)
- [x] meetings/discussions about ipfs with other orgs
  - [x] ethcore - discussed libp2p/devp2p, more future eth/ipfs interops
  - [x] ob1 - discussed using ipfs in openbazaar -- https://bitcoinmagazine.com/articles/openbazaar-integrating-interplanetary-file-system-to-help-keep-stores-open-longer-1460660998
  - [x] lantern - discussed distributed news feeds
  - [x] gitlab - discussed possible collab points (lots of good thoughts here)
  - [x] max planck - discussed collab for both resident data and archives
- [x] testing Orbit (so cool!) and reporting issues
- [x] wildcard startssl fun

**Comment by nicola** [comment]
**Summary**
I focused on understanding IPLD and the importance of separating the low level and the high level to avoid messing up combing everything together https://github.com/ipfs/specs/issues/91, lots of discussion with @jbenet and organized his coming to MIT/Berkman while in Boston. Also I wrote some tests for js-ipfs-ipld https://github.com/ipfs/js-ipfs-ipld/pull/2 https://github.com/ipfs/js-ipfs-ipld/pull/3 https://github.com/ipfs/js-ipfs-ipld/pull/5 that will probably need new revision

- [x] Writing down IPLD abstractions https://github.com/ipfs/specs/issues/91
- [x] Added tests to IPLD  https://github.com/ipfs/js-ipfs-ipld/pull/2 https://github.com/ipfs/js-ipfs-ipld/pull/3 https://github.com/ipfs/js-ipfs-ipld/pull/5
- [x] Planned Juan coming to Berkman
- [ ] preparing the presentation/interview for juan at MIT
